gazelle/python/configure.go

```diff
@@ -181,6 +182,9 @@ func (py *Configurer) Configure(c *config.Config, rel string, f *rule.File) {
 			}
 		case pythonconfig.Visibility:
 			config.AppendVisibility(strings.TrimSpace(d.Value))
+		case pythonconfig.TestFilePattern:
+			globStrings := strings.Split(strings.TrimSpace(d.Value), ",")
+			config.SetTestFilePattern(globStrings)
```

Comment:
> The empty pattern yielding no test targets is not super intuitive, IMHO.

Agreed.

That said, I would argue that the empty pattern resetting to defaults is _also_ not super intuitive. Some readers will see that and think "no test targets" while others will think "resets to default".

I like explicit things 😁. There's no ambiguity and it's easier to grok when returning to something N years down the road.

If we wanted to support yielding no test targets[^1] it should be something like `gazelle:python_test_file_pattern NONE` or have a different directive altogether (`gazelle:python_no_tests` or something).

[^1]: I can't think of a reason why someone would want that, but that doesn't mean there aren't reasons

Reply:
Yielding no test targets can (and imho should) be achieved by ignoring files with `gazelle:python_ignore_files`. That said, setting pattern to NONE should not be allowed. `gazelle:python_test_file_pattern ` (blank) should also fail.

Reply:
> yielding no test targets can (and imho should) be achieved by ignoring files with `gazelle:python_ignore_files`

I disagree. There are definitely cases where you want to yield no test targets but do want to yield lib and binary targets.

```
project/
+ BUILD.bazel
+ MODULE.bazel
+ src/myfoo/
    + BUILD.bazel
    + foo.py
    + foo_test.py
    + subdir/
        + BUILD.bazel
        + run_test.py      # py_binary
        + example_test.py  # py_library
```

Using `gazelle:python_ignore_files` would mean that _no targets_ are generated in `subdir` above, when really we'd want only py_library and py_binary targets made.

There's also the case where test files are outside of the package, as described by [PyPA Tutorial](https://packaging.python.org/en/latest/tutorials/packaging-projects/#creating-the-package-files). In that case, you can be sure that all targets in `src/mypackage` below are `py_library` and `py_binary`.

```
project/
+ BUILD.bazel
+ MODULE.bazel
+ src/mypackage/               # guaranteed to have no py_test targets
    + BUILD.bazel
    + run_electrical_test.py   # py_binary
    + electrical_test.py       # py_library
+ tests/                       # any py_* targets
    + BUILD.bazel
    + foo_test.py              # py_test
    + bar_test.py              # py_test
    + test_utils.py            # py_library
    + run_all_test.py          # py_binary
```

> also, it's a bit confusing to set pattern to `NONE`, some might thing there is `NONE` file

A valid point. Perhaps using something that is unlikely to be a filename would be better? Maybe one of these or similar?

```starlark
# gazelle:python_test_file_pattern %NONE%
# gazelle:python_test_file_pattern !!NO_FILES
# gazelle:python_test_file_pattern ~!~NO_PATTERN~!~
```

One issue is that linux filenames can be basically anything haha so it's not feasible to pick a string that _can't_ be a filename.

If we do go with a more "special" string, I'd probably want to update the `python_default_visibility` directive special strings to match.

> `gazelle:python_test_file_pattern` (blank) should also fail.

+1 to that.

Reply:
Agreed with arguments around `gazelle:python_ignore_files`. 

How about...not having any magic names at all? I mean, by the time someone wants to reset the `gazelle:python_test_file_pattern`, he/she already learned what default values, and if resetting patterns is needed, it can be done with `gazelle:python_test_file_pattern *_test,test_*`.

Reply:
I'm good with no magic names at all and a "reset" by explicitly setting the value to defaults. `gazelle:python_test_file_pattern *_test.py,test_*.py`

That said, I'd still like to mention the two (minor) issues with no magic values:

1. [can ignore] There's no way to defer to `rules_python`'s defaults. For example, if `rules_python` changes the default from `*_test.py,test_*.py` to something else, users would have to update their configs manually _if they wanted to always use the `rules_python` default_.
    + TBH this is probably a good thing because it results in less unexpected behavior.
2. We're back to the `# gazelle:python_test_file_pattern` (no value) for preventing _any_ files from mapping to `py_test`.
    + As mentioned previously, this is not very intuitive.

For (2), perhaps just making sure that's documented and tested is sufficient? Unless we implement _another_ directive `gazelle:python_no_test_files` or something (which I'm against - it seems to only complicate things).

> have dedicated no-args directive for resetting to default ... it only makes sense if applied conventionally everywhere in the plugin

+1. I'm slightly unsure about having directives to reset to default (as you said, people can just set the default manually), but if we add them they definitely should be applied everywhere.

Reply:
As far as I can see, the plugin currently has no switches for enabling/disabling _parts_ of the generation process. Setting _pattern_ to none is basically an abuse of the directive to disable tests and I'd rather have an explicit `python_no_test_files` for that so that there is clear indication no tests are being processed for a) humans, and b) plugin itself, so that it can potentially deviate its processing depending on that fact (e.g. print debug/warning message etc).

This should be, however, out of context for this PR. I'd say, we should just disallow no-arg `python_test_file_pattern`, provide no magic values and wrap-up the PR.

Have I already mentioned this is great work and nice improvement 🙂? Thank you! Looking forward to it. I have exactly the place where I will use it already.

Reply:
> Setting _pattern_ to none sounds to me like an abuse of the directive

I can see that and it's a good point.

> I'd rather have an explicit `python_no_test_files` so that there is a clear indication no tests are being processed ... we should just disallow no-arg `python_test_file_pattern`, provide no magic values and wrap-up the PR.

SGTM. I'd like to continue the discussion, so I've opened #1826.

> this is great work and nice improvement

Thanks! I've been really enjoying being able to add more features.

Reply:
All this SGTM, thanks for the discussion. I've been enjoying the read! :)